### PR TITLE
Feature/smhe 2121 overwrite functie shipmentfiles

### PR DESCRIPTION
**Pull request by natasja-n**
SMHE-2121 Overwrite functie in het inladen van shipment files
Add overwrite option for device

https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/253/

### CODE REVIEW

File: osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/installation-ws-smartmetering.xsd [hunk 1]
```diff
@@ -12,6 +12,7 @@
       <xsd:sequence>
         <xsd:element name="Device" type="tns:Device" />
         <xsd:element name="DeviceModel" type="tns:DeviceModel" />
+        <xsd:element name="overwrite" type="xsd:boolean" />
```
Comment (harrymiddelburgetall) [inline osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/installation-ws-smartmetering.xsd]: default false?

File: osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java [hunk 2]
```diff
@@ -69,23 +70,48 @@ public InstallationService() {
...
     final SmartMeter smartMeter = this.smartMeterService.convertSmartMeter(smartMeteringDevice);
+    if (existingSmartMeter.isPresent()) {
+      if (addSmartMeterRequest.getOverwrite()) {
+        smartMeter.setVersion(existingSmartMeter.get().getVersion());
+        smartMeter.setId(existingSmartMeter.get().getId());
+      } else {
+        LOGGER.error(
+            "ERROR: SmartMeter with device identification {} already exists and overwrite is not enabled",
+            messageMetadata.getDeviceIdentification());
+      }
```
Comment (harrymiddelburgetall) [inline osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java]: This is already covered by the _validateSmartMeterDoesNotExist_ method, right?

File: osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java [hunk 3]
```diff
@@ -69,23 +70,48 @@ public InstallationService() {
...
+  private void sendRequestToOsgpCore(
+      final MessageMetadata messageMetadata, final AddSmartMeterRequest addSmartMeterRequest) {
     final SmartMeteringDeviceDto requestDto =
-        this.mapperFactory.getMapperFacade().map(smartMeteringDevice, SmartMeteringDeviceDto.class);
-    this.osgpCoreRequestMessageSender.send(requestDto, messageMetadata);
+        this.mapperFactory
+            .getMapperFacade()
+            .map(addSmartMeterRequest.getDevice(), SmartMeteringDeviceDto.class);
+    requestDto.setOverwrite(addSmartMeterRequest.getOverwrite());
+    this.osgpCoreRequestMessageSender.send(requestDto, messageMetadata); // en dan??
```
Comment (harrymiddelburgetall) [inline osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java]: _en dan??_
Is there an open end to the logic?

File: osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationServiceTest.java [hunk 4]
```diff
@@ -121,18 +125,69 @@ public void addMeterProtocolInfoNotFound() {
...
+
+    doReturn(existingSmartMeter)
+        .when(this.smartMeterService)
+        .convertSmartMeter(this.smartMeteringDevice);
+
+    when(this.protocolInfoRepository.findByProtocolAndProtocolVersionAndProtocolVariant(
+            this.smartMeteringDevice.getProtocolName(),
+            this.smartMeteringDevice.getProtocolVersion(),
+            this.smartMeteringDevice.getProtocolVariant()))
+        .thenReturn(this.protocolInfo);
```
Comment (harrymiddelburgetall) [inline osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationServiceTest.java]: These lines of codes are equals with the next test except for the overwrite boolean. Did you consider a parameterized test?

File: osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java [hunk 5]
```diff
@@ -64,10 +65,28 @@ void testNonExistingSmartMeter() {
...
         Assertions.assertThrows(
             FunctionalException.class,
-            () -> this.smartMeterService.validateSmartMeterDoesNotExist(deviceIdentification));
+            () ->
+                this.smartMeterService.validateSmartMeterDoesNotExist(deviceIdentification, false));
     assertThat(exception.getExceptionType()).isEqualTo(FunctionalExceptionType.EXISTING_DEVICE);
   }
 
+  @Test
+  void testNonExistingSmartMeterWithOverwriteIsTrue() throws FunctionalException {
```
Comment (harrymiddelburgetall) [inline osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java]: This is a test with an existing smart meter. Confusing method name.

File: osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java [hunk 6]
```diff
@@ -64,10 +65,28 @@ void testNonExistingSmartMeter() {
...
+    final boolean overwrite = true;
```
Comment (harrymiddelburgetall) [inline osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java]: I'm missing a test with an existing smart meter and a overwrite false parameter

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java [hunk 7]
```diff
@@ -144,6 +147,92 @@ void addMeterNoDeviceIdentification() {
...
+        .storeAndActivateKeys(any(MessageMetadata.class), any(SmartMeteringDeviceDto.class));
+
+    installationServiceSpy.addMeter(
+        MessageMetadata.newBuilder().withCorrelationUid("123456").build(), deviceDto);
+
+    verify(this.dlmsDeviceRepository, times(1)).save(dlmsDevice);
+  }
+
+  @Test
+  void testAddMeterWithNoExistingDeviceIsNullAndOverwriteIsFalse() throws FunctionalException {
```
Comment (harrymiddelburgetall) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java]: This test only differs from the previous (testAddMeterWithNoExistingDeviceIsNullAndOverwriteIsTrue) except for the overwrite boolean. Please use a parameterized test for this.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java [hunk 8]
```diff
@@ -144,6 +147,92 @@ void addMeterNoDeviceIdentification() {
...
+        .storeAndActivateKeys(any(MessageMetadata.class), any(SmartMeteringDeviceDto.class));
+
+    installationServiceSpy.addMeter(
+        MessageMetadata.newBuilder().withCorrelationUid("123456").build(), deviceDto);
+
+    verify(this.dlmsDeviceRepository, times(1)).save(dlmsDevice);
+  }
+
+  @Test
+  void testAddMeterWithExistingDeviceAndOverwriteIsTrue() throws FunctionalException {
```
Comment (harrymiddelburgetall) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java]: If you add an boolean to the parameterized test suggested above  to indicate existing device or not you can merge the three tests.

**Review by harrymiddelburgetall — COMMENTED** [review]
Complex piece of functionality you implemented here. My comments mainly cover the tests

File: osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java [hunk 9]
```diff
@@ -69,23 +70,48 @@ public InstallationService() {
...
-
-    this.smartMeterService.validateSmartMeterDoesNotExist(deviceIdentification);
+    final Optional<SmartMeter> existingSmartMeter =
+        this.smartMeterService.validateSmartMeterDoesNotExist(
+            deviceIdentification, addSmartMeterRequest.getOverwrite());
 
     final SmartMeter smartMeter = this.smartMeterService.convertSmartMeter(smartMeteringDevice);
+    if (existingSmartMeter.isPresent()) {
+      if (addSmartMeterRequest.getOverwrite()) {
+        smartMeter.setVersion(existingSmartMeter.get().getVersion());
```
Comment (harrymiddelburgetall) [inline osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java]: Sonar states that the id field you added to the smartmeter entity is already existing in the AbstractEntity superclass. You can potentialy  avoid this construction with the hibernate merge approach.

**Review by harrymiddelburgetall — CHANGES_REQUESTED** [review]
One additional comment after looking at sonar

File: osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/installation-ws-smartmetering.xsd (hunk 1 above)
Comment (natasja-n) [reply osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/installation-ws-smartmetering.xsd]: default="false" is added.

File: osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java (hunk 2 above)
Comment (natasja-n) [reply osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java]: Right, removed duplicate code and updated method name to checkIfSmartMeterExistsAndOverwriteAllowed

File: osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java (hunk 3 above)
Comment (natasja-n) [reply osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java]: sorry, removed comment.

File: osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationServiceTest.java (hunk 4 above)
Comment (natasja-n) [reply osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationServiceTest.java]: I use the ParameterizedTest with overwrite boolean in the smart-meter-headend pr, but in some unexplainable way not in this one. They are all updated and merged now.

File: osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java (hunk 5 above)
Comment (natasja-n) [reply osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java]: Tests are updated.

File: osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java (hunk 6 above)
Comment (natasja-n) [reply osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java]: Tests are updated.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java (hunk 7 above)
Comment (natasja-n) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java]: Right, tests are updated.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java (hunk 8 above)
Comment (natasja-n) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java]: Tests are updated.

File: osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java (hunk 9 above)
Comment (natasja-n) [reply osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java]: Sonar issues are resolved. With the id and version the SmartMeter object is under control.

**Review by harrymiddelburgetall — APPROVED** [review]

**Pull request merged**